tokio/src/sync/mpsc/chan.rs

```diff
@@ -432,6 +438,27 @@ impl Semaphore for (crate::sync::semaphore_ll::Semaphore, usize) {
         self.0.add_permits(1)
     }
 
+    fn is_closed(&self) -> bool {
+        // TODO find more efficient way
+        struct NoopWaker;
+        impl crate::util::Wake for NoopWaker {
+            fn wake(self: std::sync::Arc<Self>) {}
+            fn wake_by_ref(_arc_self: &std::sync::Arc<Self>) {}
+        }
+        let waker = std::sync::Arc::new(NoopWaker);
+        let waker = crate::util::waker_ref(&waker);
+        let mut noop_cx = std::task::Context::from_waker(&*waker);
+        let mut permit = Permit::new();
+        match permit.poll_acquire(&mut noop_cx, 1, &self.0) {
+            Poll::Ready(Err(_)) => true,
+            Poll::Ready(Ok(())) => {
+                permit.release(1, &self.0);
+                false
+            }
+            Poll::Pending => false,
+        }
```

Comment:
I'm pretty sure that the semaphore's state field has a bit that we can read to check if it's closed: https://github.com/tokio-rs/tokio/blob/be7462e50fe21c3eea578e12765f501e1157072b/tokio/src/sync/semaphore_ll.rs#L116-L117

For example, we use this here:
https://github.com/tokio-rs/tokio/blob/be7462e50fe21c3eea578e12765f501e1157072b/tokio/src/sync/semaphore_ll.rs#L219-L243

I think it should be possible to have an `is_closed` method on `semaphore_ll` that looks something like this:
```rust
impl Semaphore {
    // ...

    pub(crate) fn is_closed(&self) -> bool {
        SemState(self.state.load(Acquire)).is_closed()
    }
    
    // ...
}
```

There's additional logic that occurs when closing a semaphore, to assign permits to any pending waiters prior to marking the semaphore as closed. However, if this process is currently taking place, I think it's correct to not return `true` from `is_closed` yet. So AFAICT, this implementation ought to be valid.

Reply:
I added `is_closed` method to semaphore_ll and now it is used instead of hacks with null waker.